staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -213,8 +215,72 @@ type Options struct {
 	// redundant delete API calls) and not too long that a human mistake
 	// doesn't get fixed while that human is waiting for it.
 	SyncDelay *time.Duration
+
+	// ErrorHandler will get called whenever the controller encounters
+	// a problem while trying to publish ResourceSlices. The controller
+	// will retry once the handler returns. What the handler does with
+	// that information is up to the handler. It could log the error,
+	// replace the slices if they cannot be published (see below),
+	// or force the program running the controller to fail by exiting.
+	//
+	// If some fields were dropped because the cluster does not support
+	// the feature they depend on, then the error is or wraps an
+	// [DroppedFieldsError] instance. Use [errors.As] to convert to that
+	// type:
+	//    var droppedFields *resourceslice.DroppedFieldsError
+	//    if errors.As(err, &droppedFields) { ... do something with droppedFields ... }
+	//
+	// The default is [runtime.HandleErrorWithContext] which just logs
+	// the problem.
+	ErrorHandler func(ctx context.Context, err error, msg string)
+}
+
+// DroppedFieldsError is reported through the ErrorHandler in [Options] if
+// a slice could not be published exactly as desired by the driver.
+type DroppedFieldsError struct {
+	PoolName                  string
+	SliceIndex                int
+	DesiredSlice, ActualSlice *resourceapi.ResourceSlice
+}
+
+func (err *DroppedFieldsError) Error() string {
+	// We cannot depend on go-cmp to include a diff here (not suitable for production code).
+	// The diff might be too large, too. But we can make some educated guesses....
+	disabled := err.DisabledFeatures()
+	if len(disabled) == 0 {
+		// If we get here, DisabledFeatures needs to be updated.
+		disabled = []string{"unknown"}
+	}
+	return fmt.Sprintf("pool %q, slice #%d: some fields were dropped by the apiserver, probably because these features are disabled: %s", err.PoolName, err.SliceIndex, strings.Join(disabled, " "))
+}
+
+func (err *DroppedFieldsError) DisabledFeatures() []string {
+	var disabled []string
+
+	// Both slices should have the same number of devices, but better check it.
+	for i := 0; i < len(err.DesiredSlice.Spec.Devices) && i < len(err.ActualSlice.Spec.Devices); i++ {
```

Comment:
just to be clear we're not really doing anything if one slice or the other is greater than (we would simply not enumerate through any devices for the type that has the greater number), is that what we want?

Reply:
There's nothing else that we can do: we need two of everything. But as the comment says, we should never get into a situation where storing a slice removes devices.

Reply:
Is the "need two of everything" requirement upstream from this method as well? Should this be enforced somewhere in `(c *Controller) syncPool` if so?

If we know that the lengths of `err.DesiredSlice.Spec.Devices` and `err.ActualSlice.Spec.Devices` are equal, then we can simplify to

`for i := 0; i < len(err.DesiredSlice.Spec.Devices) {`

which is a nano-optimization and maybe a little easier on the eyes

consider this a nit, not a big deal